### PR TITLE
drivers: spi: esp32: add basic SPI master support

### PR DESCRIPTION
**Pull request by sylvioalves**

### CODE REVIEW

File: soc/xtensa/esp32/Kconfig.soc [hunk 1]
```diff
@@ -16,3 +16,8 @@ config ESPTOOLPY_FLASHFREQ_80M
 	bool
 	default y
 	depends on SOC_ESP32
+
+config LOG_TIMESTAMP_SOURCE_RTOS
+	bool "Milliseconds Since Boot"
+	default y
+	depends on SOC_ESP32
```
Comment (dcpleung) [inline soc/xtensa/esp32/Kconfig.soc:23]: These do not seem to be related to this PR.

File: drivers/spi/spi_esp32_spim.c [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+	data->hal.positive_cs = 1;
+#else
+	data->hal.positive_cs = 0;
+#endif
+
+#ifdef CONFIG_SPI_ESP32_DMA
+	data->hal.dma_enabled = 1;
+#else
+	data->hal.dma_enabled = 0;
+#endif
```
Comment (dcpleung) [inline drivers/spi/spi_esp32_spim.c:186]: These `CONFIG_*` are not in `Kconfig.esp32`. Is this intentional?

File: drivers/spi/spi_esp32_spim.c [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+	if (cpha == 0) {
+		return 2;
+	}
+	return 3;
+}
+
+static void spi_esp32_get_user_settings(struct spi_esp32_data *data)
+{
+#ifdef CONFIG_SPI_ESP32_HALFDUPLEX
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.c:164]: All options in this function should be configurable by device-tree properties instead of Kconfig option. They are instance specific, so device-tree will allow to configure each instance separately.

File: drivers/spi/spi_esp32_spim.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+#define SCLK_IDX_0 HSPICLK_OUT_IDX
+#define SCLK_IDX_1 VSPICLK_OUT_IDX
+
+#define INTR_SRC_0 ETS_SPI2_INTR_SOURCE
+#define INTR_SRC_1 ETS_SPI3_INTR_SOURCE
+
+#define INST_0_ESPRESSIF_ESP32_SPI_IRQ_0 13
+#define INST_1_ESPRESSIF_ESP32_SPI_IRQ_0 17
+
+#define SPI_TIMEOUT_MSEC UINT16_MAX
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.h:23]: This seems too much for a timeout.

File: drivers/spi/spi_esp32_spim.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+#ifndef ZEPHYR_DRIVERS_SPI_ESP32_SPIM_H_
+#define ZEPHYR_DRIVERS_SPI_ESP32_SPIM_H_
+
+#define MISO_IDX_0 HSPIQ_IN_IDX
+#define MISO_IDX_1 VSPIQ_IN_IDX
+#define MOSI_IDX_0 HSPID_OUT_IDX
+#define MOSI_IDX_1 VSPID_OUT_IDX
+#define SCLK_IDX_0 HSPICLK_OUT_IDX
+#define SCLK_IDX_1 VSPICLK_OUT_IDX
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.h:15]: I would suggest to move contents of this header to `spi_esp32_spim.c` file. There is no reason to keep it separate, as nobody else uses those definitions.

File: drivers/spi/spi_esp32_spim.c [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+					 const struct spi_config *spi_cfg)
+{
+	const struct spi_esp32_config *cfg = DEV_CFG(dev);
+	struct spi_esp32_data *data = DEV_DATA(dev);
+	struct spi_context *ctx = &data->ctx;
+	spi_hal_context_t *hal = &(data->hal);
+	int freq;
+
+	/* SPI base address */
+	data->hal.hw = cfg->spi;
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.c:211]: This can be initialized during build time, from device-tree.

File: drivers/spi/spi_esp32_spim.c [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+K_TIMER_DEFINE(spi_done_timer, NULL, NULL);
+
+static inline void spi_esp32_complete(struct spi_esp32_data *data,
+				      spi_dev_t *spi, int status)
+{
+	spi_context_cs_control(&data->ctx, false);
+}
+
+static int IRAM_ATTR spi_esp32_transfer(const struct device *dev)
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.c:34]: Would be good to write some comment in commit message why `IRAM_ATTR` is used here (e.g. some hardware limitation / requirement). Just so there is some hint for people refactoring this code in future.

**Review by mniestroj — CHANGES_REQUESTED** [review]

File: drivers/spi/Kconfig.esp32 [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	depends on SOC_ESP32
+	default y
+	help
+	  Enables support for ESP32 SPI Master driver.
+
+if ESP32_SPIM
+
+config SPI_ESP32_HALFDUPLEX
+	bool "Half-duplex mode"
+	default n
```
Comment (chrta) [inline drivers/spi/Kconfig.esp32:17]: I think the "default n" lines in this file should be removed, see https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#redundant-defaults

**Review by galak — COMMENTED** [review]
Do we need the new sample or can you use tests/drivers/spi/spi_loopback ?

**Comment by sylvioalves** [comment]
> Do we need the new sample or can you use tests/drivers/spi/spi_loopback ?

No, we don't. We missed the `/tests/spi_loopback` sample.

File: drivers/spi/spi_esp32_spim.c (hunk 7 above)
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:34]: Of course, I can add this info on commit message. However, this attribute is and will be used in lots of other places as well. 
Anyway, for more information:  
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/general-notes.html#application-memory-layout

File: drivers/spi/spi_esp32_spim.c [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+	spi_hal_context_t *hal = &(data->hal);
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		spi_hal_user_start(hal);
+		spi_context_update_tx(&data->ctx, data->dfs, 1);
+	}
+
+	k_timer_start(&spi_done_timer, K_MSEC(SPI_TIMEOUT_MSEC), K_NO_WAIT);
+	while (!spi_hal_usr_is_done(hal)) {
+		if (k_timer_status_get(&spi_done_timer) > 0) {
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:46]: Will change to `k_uptime_get()` instead.

File: drivers/spi/spi_esp32_spim.c (hunk 2 above)
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:186]: No, it is not. Will fix it.

File: dts/bindings/spi/espressif,esp32-spi.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+include: spi-controller.yaml
+
+properties:
+    reg:
+      required: true
+
+    interrupts:
+      required: false
+
+    miso-pin:
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply dts/bindings/spi/espressif,esp32-spi.yaml:14]: No, I have not. We can improve later in all peripherals.

File: drivers/spi/spi_esp32_spim.c [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+		if (cpha == 0) {
+			return 0;
+		}
+		return 1;
+	}
+
+	if (cpha == 0) {
+		return 2;
+	}
+	return 3;
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:159]: No macros available.

File: drivers/spi/spi_esp32_spim.c [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+	}
+
+	ctx->config = spi_cfg;
+
+	if (ctx->config->operation & SPI_OP_MODE_SLAVE) {
+		return -ENOTSUP;
+	}
+
+	spi_esp32_configure_pin(cfg->pins.miso, cfg->signals.miso_s,
+				GPIO_INPUT);
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:230]: Will plan that in another PR.

File: drivers/spi/Kconfig.esp32 [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+	bool "ESP32 SPI dummy compensation"
+	default n
+	help
+	  Enables dummy SPI compensation cycles.
+
+config SPI_ESP32_3WIRE
+	bool "3-wire operation mode"
+	default n
+	help
+	  Enables 3-wire mode.
```
Comment (dcpleung) [inline drivers/spi/Kconfig.esp32:31]: The default for kconfig is `n` for `bool`, so there is really no need for `default n`.

Comment (sylvioalves) [reply drivers/spi/Kconfig.esp32:31]: Yes, I have fixed that already. Will push all resolved conversation soon.

File: drivers/spi/spi_esp32_spim.c [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+	struct spi_esp32_data *data = DEV_DATA(dev);
+	spi_hal_context_t *hal = &(data->hal);
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		spi_hal_user_start(hal);
+		spi_context_update_tx(&data->ctx, data->dfs, 1);
+	}
+
+	k_timer_start(&spi_done_timer, K_MSEC(SPI_TIMEOUT_MSEC), K_NO_WAIT);
+	while (!spi_hal_usr_is_done(hal)) {
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:79]: As in IDF, we can keep this nop loop for now.

**Review by chrta — CHANGES_REQUESTED** [review]
In general this looks good, i only have some minor comments. Maybe some of the can be fixed later (e.g. const for tx-buffer).

File: drivers/spi/spi_esp32_spim.c [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+
+static int IRAM_ATTR spi_esp32_transfer(const struct device *dev)
+{
+	struct spi_esp32_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	spi_hal_context_t *hal = &data->hal;
+	size_t chunk_len = spi_context_max_continuous_chunk(&data->ctx);
+
+	/* clean up and prepare SPI hal */
+	memset((uint32_t *) hal->hw->data_buf, 0, sizeof(hal->hw->data_buf));
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:65]: @chrta `hw` is `volatile`, which gives `-Wdiscarded-qualifiers` warning. Perhaps casting to `(void *)` then?

[the remark this answers is not in the document]
Comment (chrta) [reply drivers/spi/spi_esp32_spim.c:65]: No, I am ok with the current cast.

File: drivers/spi/spi_esp32_spim.c [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+	hal->send_buffer = (uint8_t *) ctx->tx_buf;
```
[the remark this answers is not in the document]
Comment (sylvioalves) [reply drivers/spi/spi_esp32_spim.c:66]: The casting was also to remove the `-Wdiscarded-qualifiers` warning (as you noted, on esp-idf the send_buffer is not `const ` .

File: drivers/spi/spi_esp32_spim.c [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+static inline uint8_t spi_esp32_get_frame_size(const struct spi_config *spi_cfg)
+{
+	uint8_t dfs = SPI_WORD_SIZE_GET(spi_cfg->operation);
+
+	dfs /= 8;
+
+	if ((dfs == 0) || (dfs > 4)) {
+		LOG_INF("Unsupported dfs, 1-byte size will be used");
```
[the remark this answers is not in the document]
Comment (chrta) [reply drivers/spi/spi_esp32_spim.c:244]: This is the only comment that is still open from me

**Review by dcpleung — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
DT bits seem reasonable

File: drivers/spi/spi_esp32_spim.c [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+							GPIO_OUTPUT);
+
+	if (ctx->config->cs == NULL) {
+		data->hal.cs_setup = 1;
+		data->hal.cs_hold = 1;
+		data->hal.cs_pin_id = 0;
+
+		spi_esp32_configure_pin(cfg->pins.csel,
+								cfg->signals.csel_s,
+								GPIO_OUTPUT | GPIO_ACTIVE_LOW);
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.c:206]: Indentation looks weird here.

File: drivers/spi/spi_esp32_spim.c [hunk 19]
```diff
@@ -0,0 +1,436 @@
...
+							cfg->signals.miso_s,
+							GPIO_INPUT);
+
+	spi_esp32_configure_pin(cfg->pins.mosi,
+							cfg->signals.mosi_s,
+							GPIO_OUTPUT_LOW);
+
+	spi_esp32_configure_pin(cfg->pins.sclk,
+							cfg->signals.sclk_s,
+							GPIO_OUTPUT);
```
Comment (mniestroj) [inline drivers/spi/spi_esp32_spim.c:197]: Some problems with indentation?

**Review by mniestroj — APPROVED** [review]

**Review by chrta — APPROVED** [review]
